service/controller/v3/resource/instance/create.go

```diff
@@ -33,187 +31,205 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 		return microerror.Mask(err)
 	}
 
-	{
-		d, err := deploymentsClient.Get(ctx, key.ClusterID(customObject), vmssDeploymentName)
-		if IsDeploymentNotFound(err) {
-			// fall through
+	if !resourceStatusExists(customObject, "Stage") {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring deployment")
+
+		computedDeployment, err := r.newDeployment(ctx, customObject, nil)
+		if controllercontext.IsInvalidContext(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "missing dispatched output values in controller context")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "did not ensure deployment")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			return nil
 		} else if err != nil {
 			return microerror.Mask(err)
 		} else {
-			s := *d.Properties.ProvisioningState
-			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deployment is in state '%s'", s))
-
-			if !key.IsFinalProvisioningState(s) {
-				r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
-				return nil
+			_, err := deploymentsClient.CreateOrUpdate(ctx, key.ClusterID(customObject), vmssDeploymentName, computedDeployment)
+			if err != nil {
+				return microerror.Mask(err)
 			}
-		}
-	}
 
-	// TODO this is a hack we have to fix as soon as the statusresource works and
-	// is enabled.
-	//
-	// Issue: https://github.com/giantswarm/giantswarm/issues/3822
-	//
-	masterVersionsValue := map[string]string{}
-	workerVersionsValue := map[string]string{}
-	{
-		var k8sClient kubernetes.Interface
-		{
-			i := key.ClusterID(customObject)
-			e := key.ClusterAPIEndpoint(customObject)
-			k8sClient, err = r.guestCluster.NewK8sClient(ctx, i, e)
+			r.logger.LogCtx(ctx, "level", "debug", "message", "ensured deployment")
+			r.logger.LogCtx(ctx, "level", "debug", "message", "setting resource status to 'Stage/DeploymentInitialized'")
+
+			err = r.setResourceStatus(customObject, "Stage", "DeploymentInitialized")
```

Comment:
Nitpick but `State` and `DeploymentInitialized ` strings are used a lot. Could they be extracted to constants?

Are they just for information to users or is there code checking for these values?

Reply:
Good shout. I added constants for them.